core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java

```diff
@@ -158,41 +173,50 @@ void schedule()
         return activeStages;
     }
 
-    private void removeCompletedStages()
+    private Set<PlanFragmentId> removeCompletedStages()
     {
         Set<StageExecution> completedStages = activeStages.stream()
                 .filter(this::isStageCompleted)
                 .collect(toImmutableSet());
         // remove completed stages outside of Java stream to prevent concurrent modification
-        completedStages.forEach(this::removeCompletedStage);
+        return completedStages.stream()
+                .flatMap(stage -> removeCompletedStage(stage).stream())
+                .collect(toImmutableSet());
     }
 
-    private void removeCompletedStage(StageExecution stage)
+    private Set<PlanFragmentId> removeCompletedStage(StageExecution stage)
     {
         // start all stages that depend on completed stage
         PlanFragmentId fragmentId = stage.getFragment().getId();
-        fragmentDependency.outgoingEdgesOf(fragmentId).stream()
+        Set<PlanFragmentId> fragmentsToExecute = fragmentDependency.outgoingEdgesOf(fragmentId).stream()
                 .map(FragmentsEdge::getTarget)
                 // filter stages that depend on completed stage only
                 .filter(dependentFragmentId -> fragmentDependency.inDegreeOf(dependentFragmentId) == 1)
-                .map(stagesByFragmentId::get)
-                .forEach(this::selectForExecution);
+                .collect(toImmutableSet());
         fragmentDependency.removeVertex(fragmentId);
         fragmentTopology.removeVertex(fragmentId);
         activeStages.remove(stage);
+        return fragmentsToExecute;
     }
 
-    private void unblockStagesWithFullOutputBuffer()
+    private Set<PlanFragmentId> unblockStagesWithFullOutputBuffer()
     {
         // find stages that are blocked on full task output buffer
         Set<PlanFragmentId> blockedFragments = activeStages.stream()
                 .filter(StageExecution::isAnyTaskBlocked)
                 .map(stage -> stage.getFragment().getId())
                 .collect(toImmutableSet());
         // start immediate downstream stages so that data can be consumed
-        blockedFragments.stream()
+        return blockedFragments.stream()
                 .flatMap(fragmentId -> fragmentTopology.outgoingEdgesOf(fragmentId).stream())
                 .map(FragmentsEdge::getTarget)
+                .collect(toImmutableSet());
+    }
+
+    private void selectForExecution(Set<PlanFragmentId> fragmentIds)
+    {
+        requireNonNull(fragmentOrdering, "fragmentOrdering is null");
+        fragmentOrdering.sortedCopy(fragmentIds).stream()
```

Comment:
why explicit copy. You can do `fragmentIds.stream().sorted(fragmentOrdering)` right?